keras/backend/jax/optimizer.py

```diff
@@ -0,0 +1,103 @@
+import jax
+from jax import numpy as jnp
+
+from keras.optimizers import base_optimizer
+
+
+class JaxOptimizer(base_optimizer.BaseOptimizer):
+    """A class for JAX specific optimizer logic.
+
+    Its purpose is to route around statelessness
+    requirements in cond ops used for EMA handling
+    and gradient accumulation handling. We do this
+    by skipping conditionals entirely.
+    """
+
+    def _backend_apply_gradients(self, grads, trainable_variables):
+        if self.gradient_accumulation_steps:
+            is_update_step = (
+                self.iterations + 1
+            ) % self.gradient_accumulation_steps == 0
+            steps = self.gradient_accumulation_steps
+
+            current_trainable_vars_value = [
+                v.value for v in trainable_variables
+            ]
+            current_optimizer_vars_value = [v.value for v in self.variables]
+
+            new_g_accs = jax.lax.cond(
+                is_update_step,
+                lambda: [
+                    jnp.zeros(x.shape, dtype=x.dtype)
+                    for x in self._accumulated_gradients
+                ],
+                lambda: [
+                    grads[i] + self._accumulated_gradients[i]
+                    for i in range(len(grads))
+                ],
+            )
+
+            grads = jax.lax.cond(
+                is_update_step,
+                lambda: [
+                    (grads[i] + self._accumulated_gradients[i]) / steps
+                    for i in range(len(grads))
+                ],
+                lambda: list(grads),
+            )
+
+            self._backend_update_step(
+                grads, trainable_variables, self.learning_rate
+            )
+            new_trainable_vars = jax.lax.cond(
+                is_update_step,
+                lambda: [v.value for v in trainable_variables],
+                lambda: current_trainable_vars_value,
+            )
+            new_opt_vars = jax.lax.cond(
+                is_update_step,
+                lambda: [v.value for v in self.variables],
+                lambda: current_optimizer_vars_value,
+            )
+
+            for value, v in zip(new_trainable_vars, trainable_variables):
+                v.assign(value)
+
+            for value, v in zip(new_opt_vars, self.variables):
+                v.assign(value)
+
+            for n_g_acc, g_acc in zip(new_g_accs, self._accumulated_gradients):
+                g_acc.assign(n_g_acc)
+
+        else:
+            self._backend_update_step(
+                grads, trainable_variables, self.learning_rate
+            )
+
+        if self.use_ema:
```

Comment:
Just curious, does the existing jax optimizer support ema? or its from the base_optimizer?

Reply:
Now JAX supports the feature -- the new unit tests check it. The previous JAX optimizer (before this PR) didn't -- only TF did.